### ISSUE TITLE
data_start is incorrect when reading an aligned file

### ISSUE THREAD

**Issue by danielocfb** [zip-rs/zip2#33]
I tried upgrading to version 1.1 and it seems as if handling of alignment is broken.

```rust
use std::env;
use std::fs::read as read_file;
use std::fs::File;
use std::io::Write as _;
use std::path::Path;

use zip::write::SimpleFileOptions;
use zip::CompressionMethod;
use zip::ZipArchive;
use zip::ZipWriter;

fn main() {
    let page_size = 4096;
    let crate_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
    let dst_dir = Path::new(&crate_dir);
    let dst_file = dst_dir.join("test.zip");

    {
        let dst_file = File::options()
            .create(true)
            .truncate(true)
            .read(false)
            .write(true)
            .open(&dst_file)
            .unwrap();

        let options = SimpleFileOptions::default()
            .compression_method(CompressionMethod::Stored)
            .with_alignment(page_size.try_into().unwrap());
        let mut zip = ZipWriter::new(dst_file);
        let contents = read_file("/usr/bin/sleep").unwrap();
        let () = zip.start_file("sleep", options).unwrap();
        let _count = zip.write(&contents).unwrap();
    }

    {
        let dst_file = File::open(dst_file).unwrap();
        let mut zip = ZipArchive::new(dst_file).unwrap();
        let file = zip.by_index(0).unwrap();
        assert_eq!(file.name(), "sleep");
        assert_eq!(file.data_start(), page_size);
    }
}
```

fails with:
```
thread 'main' panicked at src/main.rs:41:9:
assertion `left == right` failed
  left: 4092
 right: 4096
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Similar code on `0.6.4` works fine:
```rust
use std::env;
use std::fs::read as read_file;
use std::fs::File;
use std::io::Write as _;
use std::path::Path;

use zip::write::FileOptions;
use zip::CompressionMethod;
use zip::ZipArchive;
use zip::ZipWriter;

fn main() {
    let page_size = 4096;
    let crate_dir = env::var("CARGO_MANIFEST_DIR").unwrap();
    let dst_dir = Path::new(&crate_dir);
    let dst_file = dst_dir.join("test.zip");

    {
        let dst_file = File::options()
            .create(true)
            .truncate(true)
            .read(false)
            .write(true)
            .open(&dst_file)
            .unwrap();

        let options = FileOptions::default().compression_method(CompressionMethod::Stored);
        let mut zip = ZipWriter::new(dst_file);
        let contents = read_file("/usr/bin/sleep").unwrap();
        let _align = zip
            .start_file_aligned(
                "sleep",
                options,
                page_size.try_into().unwrap(),
            )
            .unwrap();
        let _count = zip.write(&contents).unwrap();
    }

    {
        let dst_file = File::open(dst_file).unwrap();
        let mut zip = ZipArchive::new(dst_file).unwrap();
        let file = zip.by_index(0).unwrap();
        assert_eq!(file.name(), "sleep");
        assert_eq!(file.data_start(), page_size);
    }
}
```

Am I misunderstanding what is being aligned?

**Comment by d-e-s-o** [comment]
This exact test case was broken *again* with 2.1.4. Please reopen.

```
thread 'main' panicked at src/main.rs:41:9:
assertion `left == right` failed
  left: 35
 right: 4096
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I would bisect it, but your history is utterly broken.
```
The first bad commit could be any of:
3436a569ab1f76944f45f2dfa0eab5c8aaeec6c1
aa896ce9927f591850ab77fbc9146aa42fc550ec
7b2e31ca7fcff012d77f79bb3ada3a06bfa07d0c
bf3e7510a6a39dcafd5d99df5d954b8df72b2460
0d9fd0da0b25898c1f995e811bdedd8ed2025d79
7f641d23a32da92b0fc5203cff935baf45591785
0f0ef1dd960a5410f013160a946ac576b282fabb
9e12b6db0cb7958fb5abc4169101ff5757ee073c
0ad79e9ab3dccbcfc6112c0e4948f665478edc44
40d2da84f753be243795d2d03340aac6ddc1990b
19e5af84064b959f19377b88a0b7e6e768d81c7c
60f7c637e6b820b023fefa24442e67cb743b8a5a
32f946e78851615ac445579db427241cefdf5e7f
```

Would it be possible to have some kind of quality assurance by any chance?

**Comment by Pr0methean** [comment]
Could you please explain how the use case in which you're observing the problematic behavior is different from what's tested by the method `test_alignment()` added to `write.rs` in https://github.com/zip-rs/zip2/commit/c3d9123abf43795659df9b3bb247622f0b0b1d67? I'd intended that test to catch any regression of this bug.

**Comment by d-e-s-o** [comment]
The following test reproduces it:
```rust
    #[test]
    fn test_alignment() {
        let page_size = 4096;
        let mut data = Vec::new();
        {
            let options = SimpleFileOptions::default()
                .compression_method(CompressionMethod::Stored)
                .with_alignment(page_size);
            let mut zip = ZipWriter::new(io::Cursor::new(&mut data));
            let contents = b"sleeping";
            let () = zip.start_file("sleep", options).unwrap();
            let _count = zip.write(&contents[..]).unwrap();
        }

        {
            let mut zip = ZipArchive::new(io::Cursor::new(&mut data)).unwrap();
            let file = zip.by_index(0).unwrap();
            assert_eq!(file.name(), "sleep");
            assert_eq!(file.data_start(), page_size.into());
        }
    }
```
The difference is that it doesn't use `finish_into_readable` method. I don't know why, with that method, the bug isn't hit.

**Comment by Pr0methean** [comment]
Ah, this is actually a different bug: the data starts at byte 4096 as expected (I confirmed this by adding `assert_eq!(data[4096..4104], b"sleeping"[..]);` to the test case you provided, but `data_start()` incorrectly returns 35 when the archive is read back in.

**Comment by d-e-s-o** [comment]
Makes sense. But irrespective of that the zip archive created with `2.1.4` or more recent is still broken compared to previous versions. That is without any involvement of `data_start()`.

Here are two archives created using `zip` with basically the above logic (no compression, page alignment etc.), using different versions of `zip`:
[test-2.1.3.zip](https://github.com/user-attachments/files/16367270/test-2.1.3.zip)
[test-2.1.5.zip](https://github.com/user-attachments/files/16367271/test-2.1.5.zip)

`test-2.1.5.zip` can't be unpacked properly:
```sh
$ unzip test-2.1.5.zip
Archive:  test-2.1.5.zip
 extracting: test-stable-addrs-stripped-elf-with-dwarf.bin   bad CRC 2886939d  (should be d9becfd9)
 extracting: zip-dir/test-no-debug.bin   bad CRC 239d2efa  (should be 0e0e8dc5)
 extracting: libtest-so.so            bad CRC ac70132d  (should be f886cd80)
 extracting: libtest-so-no-separate-code.so   bad CRC 82120180  (should be 33423096)
```

**Comment by Pr0methean** [comment]
That's yet another different issue: #195.

**Comment by d-e-s-o** [comment]
Yeah, I understand. I am not blocked on this getting fixed as I've blacklisted 2.1.4 and higher and I don't really use the crate in production. That said, it is very annoying to have to deal with regressions like this on a 1.x or 2.x crate. Almost equally bad is not having a buildable commit history. Root causing issues like this should be trivial and instead we are running in circles for nothing.

> That's yet another different issue: #195.

It's possible, but I kind of doubt it. The issue may manifest as checksum mismatches in my paste, but I am not convinced that is necessarily the actual issue. For one, #195 mentions version 2.1.3. This version is definitely behaving properly for my use case. Based on my limited testing, data offsets in the local file headers seemed off (though again, that may not be the actual root cause either, it is hard for me to tell).

If we compare the hex dumps there are a bunch of differences between the two, most notably towards the end. I am not fluent enough to judge validity of those differences at this point.

![screenshot-2024-07-24_14-40-55](https://github.com/user-attachments/assets/a9ffaff9-5b76-4317-99ed-5774e43f0500)

**Comment by Pr0methean** [comment]
I've mentioned my need for a co-admin on the relevant Slack channels at work; hopefully someone can either recommend one, or step up to the plate themselves and be at least as qualified as me.

**Comment by danielocfb** [comment]
> > That's yet another different issue: #195.
> 
> It's possible, but I kind of doubt it.

Also, I am not setting `large_file(true)`.

**Comment by Pr0methean** [comment]
#221 may fix this; if so, #228 will be able to verify the fix after being rebased.

**Comment by d-e-s-o** [comment]
> On my machine #221 fixes this, and #228 will be able to verify the fix after being rebased.

I can confirm that I no longer see issues at 6d8ab6224b78535d9d71e54911a518265f1d42d8. Thanks!
